[IBC] Make the IBC Host a submodule with access to the bus

h5law opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 13 Jul 23 10:41 UTC
This pull request includes several changes across multiple files. Here is a summary of the changes:

1. In the file `ibc/handle_message_test.go`, the file was renamed to `ibc/ibc_msg_mempool_test.go` with a similarity index of 85%.
2. In the same file, the imports were modified, and the `strings` package was added.
3. The test function `TestHandleMessage_ErrorAlreadyInMempool` was renamed to `TestEmitMessage_MessageAddedToLocalMempool`. Some code within the renamed test function was also modified, including changes in the preparation of test data and the addition of a transaction to the mempool.
4. The test function `TestHandleMessage_BasicValidation_Message` was renamed to `TestIBCMessage_BasicValidation_Message`.
5. The test function `TestHandleMessage_BasicValidation_Transaction` was renamed to `TestIBCMessage_BasicValidation_Transaction`.
6. A new test function `TestHandleMessage_ErrorAlreadyInMempool` was added to check for the error of having a duplicate transaction in the mempool.
7. A new test function `TestHandleMessage_ErrorAlreadyCommitted` was added to check for the error of having an already committed transaction.
8. The test function `TestHandleMessage_GetIndexedMessage` was modified to include changes in the preparation of the environment.
9. The test function `TestHandleMessage_AddToMempool` was removed.

Additionally, other files such as `treestore_module.go`, `emitter.go`, `submodule.go`, `config.validator4.json`, `persistence/ibc.go`, `defaults.go`, `bus_module.go`, `ibc_store_module.go`, `bulk_store_cache.go`, `ibc_host_module.go`, `config.validator4.json`, `persistence/test/benchmark_state_test.go`, `p2p/README.md`, `runtime/manager_test.go`, `shared/node.go`, `persistence/test/manager_test.go`, `shared/modules/bulk_store_cache.go`, `ibc.go`, `shared/node.go`, `ics24.md`, `main_test.go` have also been modified.

Please let me know if you need more information or details about any specific change.
<!-- reviewpad:summarize:end -->

## Issue

Fixes #854 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Move IBC host into its own submodule
- Enable local ProvableStore instances to emit events to alter the IBC state tree

## Testing

- [x] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made


## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

bryanchriswhite commented on ibc/host/module.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func (h *ibcHost) GetModuleName() string  { return modules.IBCHostModuleName }
+func (h *ibcHost) GetBus() modules.Bus    { return h.bus }
+func (h *ibcHost) SetBus(bus modules.Bus) { h.bus = bus }
```
This is redundant because you've already embedded the `Ingegrable` base module in `ibcHost`.

```suggestion
```

bryanchriswhite commented on internal/testutil/ibc/mock.go:
```diff
@@ -10,34 +10,31 @@ import (
 
 // BaseIBCMock returns a mock IBC module without a Host
-func BaseIBCMock(t gocuke.TestingT, busMock *mockModules.MockBus) *mockModules.MockIBCModule {
+func BaseIBCMock(t gocuke.TestingT, bus modules.Bus) *mockModules.MockIBCModule {
```
Can you speak to your experience regarding the parameter type when passing a bus argument? I don't have an informed enough opinion yet and it seems like you've encountered something here.

h5law replied:
This is for use in the consensus e2e tests it uses a real bus not a mock so the arg must be a real one as this is where its being used currently

bryanchriswhite replied:
I was looking for more about the "why" bus over mock bus, what are the constraints / requirements that drive your decision here?

h5law replied:
So if we look in `consensus/e2e_tests/utils_test.go` at the function `CreateTestConsensusPocketNode()` it is passed a real bus. As such it gives a real bus to each of the mocks it creates. I cannot use the real mock as a bus mock in this base function call and as its the only place using it, for now I thought its best to just follow the pattern and use the real bus its given.

In reality the IBC module is mostly a shepherd for the IBC host, I think any real mock that uses the IBC module will probably just need it to create the IBC host (or another submodule later down the line for servicers to handle IBC relaying). These submodules are registered to the bus, and are only retrievalbe from the bus - so a real bus makes sense here?

bryanchriswhite commented on ibc/emit_message_test.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+func TestEmitMessage_MessageAddedToLocalMempool(t *testing.T) {
+	t.Skip()
```
Good habit :+1:

bryanchriswhite commented on ibc/main_test.go:
```diff
@@ -73,6 +91,17 @@ func prepareEnvironment(
+	err = testConsensusMod.Start()
+	require.NoError(t, err)
+
+	// TODO: I want to mock this as using a real p2p module is hard to test
```
Can you elaborate on what you mean here? I wasn't able to follow.

Consider using a specific [TODO type](https://github.com/pokt-network/pocket/blob/main/Makefile#L448).

h5law replied:
This was a personal comment for myself ahah. I think this comment is removed when the P2P mock comes in.

bryanchriswhite commented on ibc/main_test.go:
```diff
@@ -73,6 +91,17 @@ func prepareEnvironment(
+	testP2PMod := newTestP2PModule(bus)
+	err = testP2PMod.Start()
+	require.NoError(t, err)
+	**/
```
nit: We should prefer not to commit commented code because it's much more difficult to determine when it becomes obsolete.

I realize this PR is still marked as a draft but I'm reviewing the same, since you explicitly asked for review. :sweat_smile:

h5law replied:
Yeah I had to commit it as me and @Olshansk did some work on the branch together the other day to get the P2P mock working in the test. Unsure whether I should rebase a WIP or not after the other day o-o

bryanchriswhite commented on internal/testutil/ibc/mock.go:
```diff
@@ -10,34 +10,31 @@ import (
-func IBCMockWithHost(t gocuke.TestingT, _ modules.EventsChannel) *mockModules.MockIBCModule {
+// BaseIBCHostMock returns a mock IBC Host submodule
+func BaseIBCHostMock(t gocuke.TestingT, busMock *mockModules.MockBus) *mockModules.MockIBCHostModule {
 	ctrl := gomock.NewController(t)
```
(see: https://github.com/pokt-network/pocket/pull/868/commits/36d4d9d06899d8b45ba8da5a1a0d0452342b66af#r1252166693)

```suggestion
	t.Helper()

	ctrl := gomock.NewController(t)
```

bryanchriswhite commented on internal/testutil/ibc/mock.go:
```diff
@@ -10,34 +10,31 @@ import (
 
-// IBCMockWithHost returns a mock IBC module with a Host
-func IBCMockWithHost(t gocuke.TestingT, _ modules.EventsChannel) *mockModules.MockIBCModule {
+// BaseIBCHostMock returns a mock IBC Host submodule
+func BaseIBCHostMock(t gocuke.TestingT, busMock *mockModules.MockBus) *mockModules.MockIBCHostModule {
```
:100: :ok_hand:

bryanchriswhite commented on shared/modules/ibc_host_module.go:
```diff
@@ -0,0 +1,197 @@
+// are also responsible for receiving any IBC packets from another chain and verifying them through
+// the light clients they manage
+// https://github.com/cosmos/ibc/tree/main/spec/core/ics-024-host-requirements
+type IBCHostModule interface {
```
How would you feel about `IBCHostSubmodule` instead?

I'm thinking back to the discussion we had earlier about naming the implementation's file `submodule.go` over `module.go`. I acknowledge that here w're in the `modules` package but I would argue this is for 2 reasons, to avoid import cycles and reduce complexity.

I imagine the number of submodules eventually, if not quickly, outnumbering modules. I anticipate the effects of this will include cluttering up the bus module interface with getter functions and accumulating more interface files in this package. By differentiating the naming convention of submodule interfaces it should become much easier to make the distinction between them just from the name

bryanchriswhite commented on shared/modules/ibc_host_module.go:
```diff
@@ -0,0 +1,197 @@
+	GetTimestamp() uint64
+}
+
+// TODO: Split into mutliple interfaces per ICS component and embed in the handler
```
(nit: consider using a specific TODO type)

bryanchriswhite commented on ibc/handle_message_test.go:
```diff
@@ -286,6 +287,9 @@ func TestHandleMessage_AddToMempool(t *testing.T) {
 func prepareUpdateMessage(t *testing.T, key, value []byte) (*ibcTypes.IBCMessage, *coreTypes.Transaction) {
 	t.Helper()
 	msg := ibcTypes.CreateUpdateStoreMessage(key, value)
+	letter, err := messaging.PackMessage(msg)
+	require.NoError(t, err)
+	t.Log(letter.GetContent().GetTypeUrl())
```
nit:

```suggestion
	poktEnvelope, err := messaging.PackMessage(msg)
	require.NoError(t, err)
	t.Log(poktEnvelope.GetContent().GetTypeUrl())
```

dylanlott commented on ibc/module.go:
```diff
@@ -1,34 +1,28 @@
 	cfg    *configs.IBCConfig
 	logger *modules.Logger
 
-	// Only a single host is allowed at a time
```
Does this comment need to be recaptured somewhere else or is it no longer applicable?

dylanlott commented on ibc/module.go:
```diff
@@ -69,91 +65,36 @@ func (m *ibcModule) Start() error {
 }
 
 func (m *ibcModule) Stop() error {
+	m.logger.Info().Msg("🛑 Stopping IBC Module 🛑")
```
🛑 💯

Olshansk replied:
We should do this everywhere

bryanchriswhite replied:
Worth including in [the docs](https://github.com/pokt-network/pocket/blob/main/shared/modules/doc/README.md#stop-the-module)?

dylanlott commented on ibc/store/emmitter.go:
`emmitter.go` --> `emitter.go`

dylanlott commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,187 @@
+)
+
+var (
+	_         modules.BulkStoreCacher = &bulkStoreCache{}
+	cacheDirs                         = func(storesDir string) string { return fmt.Sprintf("%s/caches", storesDir) }
```
Nit: this space is super weird to read, consider spacing them out to avoid the formatting change or making them individual var declarations instead of grouping them together.

h5law commented:
🚨 I REWROTE HISTORY 🚨

As I took this branch out of a WIP draft I have rebased and cleaned up the commit history (fixups etc). This is the only time this will happen as this PR is now ready for review.

CC: @Olshansk @bryanchriswhite

Olshansk commented on ibc/docs/ics24.md:
```diff
@@ -177,15 +162,29 @@ See: [PROTOCOL_STATE_HASH.md](../../persistence/docs/PROTOCOL_STATE_HASH.md#ibc-
 ### Caching
 
-Every local change made to the IBC store (`update`/`delete`) is stored in an in-memory cache. These caches are written to disk by the [`StoreManager`](../store/store_manager.go).
+Every local change made to the IBC store (`update`/`delete`) is stored in an in-memory cache. These caches are written to disk by the `BulkStoreCacher`, upon the receipt of a new height event from consensus. This means that the cache for any given height is always written to disk all at once.
```
```suggestion
Every local change made to the IBC store (`update`/`delete`) is stored in an in-memory cache. These caches are written (i.e. flushed) to disk by the `BulkStoreCacher`, upon the receipt of a new height event from consensus. This means that the cache for any given height is always written to disk all at once.
```

Olshansk commented on ibc/docs/ics24.md:
```diff
@@ -177,15 +162,29 @@ See: [PROTOCOL_STATE_HASH.md](../../persistence/docs/PROTOCOL_STATE_HASH.md#ibc-
-Every local change made to the IBC store (`update`/`delete`) is stored in an in-memory cache. These caches are written to disk by the [`StoreManager`](../store/store_manager.go).
+Every local change made to the IBC store (`update`/`delete`) is stored in an in-memory cache. These caches are written to disk by the `BulkStoreCacher`, upon the receipt of a new height event from consensus. This means that the cache for any given height is always written to disk all at once.
+
+The `BulkStoreCacher` also keeps track of the maximum height stored in the cache. The `max_height_stored` value (default `5`) is used when a new height event comes in to prune old entries in the cache stored on disk.
```
```suggestion
The `BulkStoreCacher` also keeps track of the maximum height stored in the cache. The `max_height_stored` value (default `5`) is used when a new height event comes in to prune old entries in the cache stored on disk in a "circular buffer" like fashion.
```

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+)
+
+// bulkStoreCache holds an in-memory map of all the provable stores in use
+// RESEARCH: Look into parrelising the caching methods
```
Do you mean parallelizing with a `z`?

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+type bulkStoreCache struct {
+	base_modules.IntegrableModule
+
+	m sync.Mutex
```
Add a comment on what this is for. If it's a mutext for the `stores` map, group them together inside the struct

Olshansk commented on ibc/module.go:
```diff
@@ -69,91 +63,67 @@ func (m *ibcModule) Start() error {
+			break
+		}
+		pruneHeight := newHeight - m.cfg.Host.BulkStoreCacher.MaxHeightStored
+		if err := bsc.PruneCaches(pruneHeight); err != nil {
```
Isn't this always going to prune small heights?

E.g. If new height is `102` and max height stores is `100`, we'll be running height 2?

I feel like you're attempt is having a circular buffer in the cache but I'm not convinced that's what it's actually doing right now.

h5law replied:
So what this does is we are at height 100, we have max stored height of 5, so we have stored 95-100 heights in our cache. when 101 comes in we prune 95 and now store 96-101. The cache doesnt need to be very large just enough to be able to replay if we feel the tx didnt propagate. max height stored is customisable but defaults to 5

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+	s.m.Lock()
+	defer s.m.Unlock()
+	if _, ok := s.stores[name]; ok {
+		return coreTypes.ErrIBCStoreAlreadyExists(name)
```
Nice 👌

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+func (s *bulkStoreCache) GetAllStores() map[string]modules.ProvableStore {
+	s.m.Lock()
+	defer s.m.Unlock()
+	stores := make(map[string]modules.ProvableStore, len(s.stores))
+	for name, store := range s.stores {
+		stores[name] = store
+	}
+	return stores
```
Why not just return `s.stores`?

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+		return err
+	}
+	for _, store := range s.stores {
+		if err := store.FlushEntries(disk); err != nil {
```
Consider adding a log line for each store flushed (with a reference/id as to which one it is) so we can track it easily in case a single one fails.

Potentially just on error.

Ditto below for pruning.

Olshansk commented on ibc/store/bulk_store_cache.go:
```diff
@@ -0,0 +1,188 @@
+	if dir == ":memory:" {
+		return kvstore.NewMemKVStore(), nil
+	}
+	store, err := kvstore.NewKVStore(cacheDirs(dir))
+	if err != nil {
+		return nil, err
+	}
+	return store, nil
```
```suggestion
	return kvstore.NewKVStore(cacheDirs(dir))
```

Olshansk commented on shared/node.go:
```diff
@@ -183,14 +183,17 @@ func (node *Node) handleEvent(message *messaging.PocketEnvelope) error {
 		return node.handleDebugMessage(message)
 	case messaging.ConsensusNewHeightEventType:
-		return node.GetBus().GetP2PModule().HandleEvent(message.Content)
+		if err := node.GetBus().GetP2PModule().HandleEvent(message.Content); err != nil {
```
Look at `StateMachineTransitionEventType` and follow th same pattern leveraging `multierr.Combine`

Olshansk commented on shared/modules/ibc_store_module.go:
```diff
@@ -0,0 +1,50 @@
+	GetStore(name string) (ProvableStore, error)
+	RemoveStore(name string) error
+	GetAllStores() map[string]ProvableStore
+	FlushAllEntries() error
```
OPTIONAL: `FlushEntriesFromCachestoStores`

I realize this is a bit pedantic but just TOL (thinking out loud).

@bryanchriswhite wdyt?

bryanchriswhite replied:
I like the direction, what about just `FlushCachesToStore`?

Assuming no understanding of what I'm looking at (which is pretty accurate at this point :sweat_smile:), I like `FlushEntriesFromCachesToStores` better than `FlushAllEntries` as it's not clear to me from the interface definition what an "entry" is and/or how it relates to anything else described in the interface. I think `FlushCachesToStore` better captures the minimal relevant context.

Olshansk commented on shared/modules/ibc_host_module.go:
```diff
@@ -0,0 +1,205 @@
+// the host is responsible for managing the IBC state and interacting with consensus in order for
+// any IBC packets to be sent to another host on a different chain (via an IBC relayer). The hosts
+// are also responsible for receiving any IBC packets from another chain and verifying them through
+// the light clients they manage
```
```suggestion
// the IBC light clients they manage
```

Olshansk commented on runtime/configs/proto/ibc_config.proto:
```diff
@@ -12,6 +12,15 @@ message IBCConfig {
+}
+
+message BulkStoreCacherConfig {
+    uint64 max_height_stored = 1;      // maximum number of heights to be stored in the cache before pruning
```
I left another comment around how this is implemented. You say this is the `maximum number of heights` but it is actually the ACTUAL height during the height check.

I think there's a descpenecy/issue/confusion in tbothe documentation & implementation.

1. Please update both
2. If there aren't any tests, please add them
3. If it is meant to be the `number`, make the comment say it's a "circular buffer" like object.

I'm going to put a bit more though into proper naming here.

Olshansk commented on runtime/configs/proto/ibc_config.proto:
```diff
@@ -12,6 +12,15 @@ message IBCConfig {
+    BulkStoreCacherConfig bulk_store_cacher = 3;
+}
+
+message BulkStoreCacherConfig {
```
Add an `INVESTIGATE: Should we cap the cache size by size (bytes) instead of heights which could, theoretically, grow unbounded`

Olshansk commented on persistence/test/ibc_test.go:
```diff
@@ -0,0 +1,122 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestSetIBCStoreEntry(t *testing.T) {
```
Please follow the same pattern elsewhere

```suggestion
func TestIBC_SetStoreEntry(t *testing.T) {
```

Olshansk commented on internal/testutil/ibc/mock.go:
```diff
@@ -10,34 +10,72 @@ import (
+}
+
+// BaseBulkStoreCacherMock returns a mock BulkStoreCacher submodule mock
+func BaseBulkStoreCacherMock(t gocuke.TestingT, bus modules.Bus) *mockModules.MockBulkStoreCacher {
```
Do we need to publically expose all these mock helpers?

Olshansk commented on ibc/store/provable_store.go:
```diff
@@ -37,20 +41,23 @@ func (c *cachedEntry) prepare() (key, value []byte) {
-	prefix coreTypes.CommitmentPrefix // []byte(name)
-	cache  map[string]*cachedEntry    // in-memory cache of local changes to be written to disk
+	m          sync.Mutex
+	bus        modules.Bus                // used to interact with persistence (passed from IBCHost)
```
Order for clarity.

1. Bus at top
2. Mutex next to the object (i.e. cache?) it's responsible for protecting
3. Add comment to the utex

Olshansk commented on ibc/main_test.go:
```diff
@@ -28,26 +32,63 @@ func TestMain(m *testing.M) {
+	p2pMock.EXPECT().
+		Broadcast(gomock.Any()).
+		Do(func(msg *anypb.Any) {
+		}).
```
If you're not going to `Do` anything, avoid including it

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+				require.Equal(t, err, tc.expectedError)
+				return
+			}
+			require.Equal(t, value, tc.expectedValue)
```
Add. a `require.NoError` in this case

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+				require.Equal(t, err, tc.expectedError)
+				return
+			}
+			require.Equal(t, value, tc.expectedValue)
```
Add a `require.NoError`

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+	require.NoError(t, err)
+	require.Len(t, keys, 3)
+	for i, key := range keys {
+		require.Equal(t, string(key), "test/1/test/"+string(kvs[i].key))
```
Can you move `"test/1/test/"+string(kvs[i].key)` into a var and explain how you got it?

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+	require.NoError(t, provableStore.FlushEntries(cache))
+	keys, values, err := cache.GetAll([]byte{}, false)
+	require.NoError(t, err)
+	require.Len(t, keys, 3)
```
Add a comment on why `3`

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+	return p2pMock
+}
+
+func newTestRuntimeConfig(t *testing.T) *runtime.Manager {
```
Add a `// TECHDEBT` comment to centralize these types of helpers in the future.

cc @adshmh and @bryanchriswhite just for context & visibility

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+	require.NoError(t, cache.Stop())
+}
+
+func TestProvableStore_PruneCache(t *testing.T) {
```
Per my other comments, can you point me to where this tests a real circular buffer? That we reach the "max number" and no more?

h5law replied:
This test is for the single provable store that logic is handled by the bulk store cacher will be adding the test shortly

Olshansk commented on ibc/store/provable_store_test.go:
```diff
@@ -0,0 +1,540 @@
+	p2pMock.EXPECT().GetBus().Return(bus).AnyTimes()
+	p2pMock.EXPECT().
+		Broadcast(gomock.Any()).
+		Do(func(msg *anypb.Any) {
```
Don't have `Do` if you don't need it. Ditto elsewhere.

h5law commented on build/config/config.validator1.json:
```diff
@@ -60,7 +60,9 @@
-    "stores_dir": "/var/ibc"
+    "host": {
+        "private_key": "0ca1a40ddecdab4f5b04fa0bfed1d235beaa2b8082e7554425607516f0862075dfe357de55649e6d2ce889acf15eb77e94ab3c5756fe46d3c7538d37f27f115e",
+        "stores_dir": "/var/ibc"
```
hmm I put it in host as the ibc module doesn't really use the stores itself it goes thru the host. potentially in the future when servicers have a IBC module flow for relaying they might need it too so will change now save time later

h5law commented:
@Olshansk added tests to cover the flushing and pruning works on new height events

Olshansk commented on ibc/docs/ics24.md:
````diff
@@ -127,42 +135,19 @@ _Note: In both types described above the `key` field **must** already be prefixe
-  Handler--Transaction-->Mempool
-  M1--Transaction-->M2
-```
+1. Create the relevent `IBCMessage` type
````
@h5law If you don't do it in this PR, can you add a TODO for it?

h5law replied:
@Olshansk my bad missed this one, will do now

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+	require.NoError(t, cache.Stop())
+}
+
+// using MaxStoredHeight = 3
```
Make sure to update comments too when you rename the variable

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+	require.NoError(t, cache.Stop())
+}
+
+// using MaxStoredHeight = 3
```
Where is this set? Is there any way you can set it explicitly (via code) in the test so, for example, we could test other values too?

h5law replied:
`ibc/main_test.go` in the runtime setup function as its a config var

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+
+	require.NoError(t, cache.Stop())
+
+	// flush the cache
```
You only flush AFTER you've cleared everything.

Do you not want to:

1. Store
2. Flush
3. Restore
4. Validate (non empty)

5. Clear

6. Flush
7. Restore
8. Validate (empty)

h5law replied:
Yeah I am flushing to confirm the original flush got everything. So I take the first flush, confirm it flushed the correct amount, clear the disk cache and then flush again without any updates to ensure nothing was left. Your method would require I restore and flush again as you cannot access the cache in memory

Olshansk replied:
> clear the disk cache and then flush again without any updates to ensure nothing was left

So is this part of the test or is it a "cleanup" task. If it's a cleanup, use `t.cleanup` (and update other parts of the code as well). This way it's clear that we're not testing flushing, but simply have tests that are dependant on each other.

h5law replied:
Its not cleanup, its ensuring everything was cached the first time round. cleanup is just stopping the store and main_test.go deletes the directory when done

h5law replied:
If the second flush put anything in the cache we would know the first cache job didnt work correctly its a double verification

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+
+	testCases := []struct {
+		name            string
+		heights         []uint64
```
Can you make this more explitict or add a comment?

1. `heightsStored` or `heights [] uint64 // the heights that have been committed to the state`

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+	testCases := []struct {
+		name            string
+		heights         []uint64
+		expectedHeights []uint64
```
Similar to the above. Is this `heightsCached`?

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+		name            string
+		heights         []uint64
+		expectedHeights []uint64
+		length          int
```
What does the length represent? Rename or add a comment

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+			// set height
+			publishNewHeightEvent(t, ibcMod.GetBus(), tc.heights[0])
+
+			for _, height := range tc.heights[1:] {
```
Insert every key at every height?

h5law replied:
Yeah

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+			require.Len(t, keys, tc.length)
+			require.Len(t, values, tc.length)
+
+			// iterate over the expected heights and check the keys and values are correct
```
By expected, do you mean `expect to be cached` and by correct do you mean `in the cache`?

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+						require.Equal(t, values[j], []byte{})
+						continue
+					}
+					require.Equal(t, values[j], kvs[j%4].value)
```
Where do we validate that the previous heights are no longer cached?

h5law replied:
added comment for clarity this function actually does this validation

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,242 @@
+
+			newKeys, newValues, err := cache.GetAll([]byte{}, false)
+			require.NoError(t, err)
+			require.Len(t, newKeys, 0)
```
What does this extra check (after clearing everything) provide in the context of this test?

Olshansk reviewed (changes requested):
@h5law I have some questions around the test. It might just be me but its not completely clear how the test guarantees what we expect.

@bryanchriswhite Do you mind taking a look at `ibc/ibc_handle_event_test.go` as well

Olshansk commented on ibc/docs/ics24.md:
````diff
@@ -147,6 +147,38 @@ Upon making a local change to the IBC store the host will:
 4. Add the `Transaction` to the local mempool
 5. Broadcast the `Transaction` via the `P2P` module for the other nodes to include it in their mempools
 
+```mermaid
````
🔥

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -94,7 +94,7 @@ func TestHandleEvent_FlushCaches(t *testing.T) {
 }
 
-// using MaxStoredHeight = 3
+// using MaxHeightCached = 3
```
```suggestion
// using MaxHeightCached = 3; set in `ibc/main_test.go`
```

Olshansk commented on ibc/ibc_handle_event_test.go:
```diff
@@ -125,49 +125,49 @@ func TestHandleEvent_PruneCaches(t *testing.T) {
+		name                  string
+		heightsStored         []uint64 // the different heights where entries are written to the cache
+		expectedHeightsCached []uint64 // the different heights expected in the cache after pruning
+		cacheLength           int      // the length of the cache after pruning
```
```suggestion
		cacheSize           int      // the number of items in the cache after pruning
```

Olshansk reviewed (changes requested):
A couple more nits.

PTAL at https://github.com/pokt-network/pocket/pull/868#discussion_r1260269000 as well

Olshansk reviewed (approved):
@h5law Please be cognizant of the branches (based branches, merge order etc...), since I just approved two of them.

I suggest:

1. Merge the base to main
2. Merge this one with main
3. Push this one up + update base to main
4. Sanity check this PR is good
5. Merge this into main

Other than that, 🚢

bryanchriswhite commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,220 @@
+
+func TestHandleEvent_FlushCaches(t *testing.T) {
+	mgr, _, _, _, ibcMod := prepareEnvironment(t, 1, 0, 0, 0)
+	tmpDir := mgr.GetConfig().IBC.StoresDir + "/caches"
```
nit: prefer using the `path/filepath` over hardcoding os-dependent path delimiters (`/`: *nix, `\` windows)

```suggestion
	tmpDir := filepath.Join(mgr.GetConfig().IBC.StoresDir, "caches")
```

bryanchriswhite commented on ibc/ibc_handle_event_test.go:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	for _, kv := range kvs {
+		if kv.value != nil {
+			require.NoError(t, store.Set(kv.key, kv.value))
+		} else {
+			require.NoError(t, store.Delete(kv.key))
+		}
```
Could you add a comment here to explain the rationale behind logic?
